packages/opentelemetry-basic-tracer/src/BasicTracer.ts

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as types from '@opentelemetry/types';
+import { BaseScopeManager } from '@opentelemetry/scope-base';
+import {
+  ALWAYS_SAMPLER,
+  NoopLogger,
+  NoopSpan,
+  randomSpanId,
+  randomTraceId,
+} from '@opentelemetry/core';
+import { BasicTracerConfig } from '../src/types';
+
+/**
+ * This class represents a basic tracer.
+ */
+export class BasicTracer implements types.Tracer {
+  static INVALID_ID = '0';
+
+  private defaultAttributes: types.Attributes;
+  private logger: types.Logger;
+  private sampler: types.Sampler;
+  private scopeManager?: BaseScopeManager;
+
+  // TODO: consume invalid span context from `SpanContext.INVALID`
+  static defaultSpan = new NoopSpan({
+    traceId: BasicTracer.INVALID_ID,
+    // TODO: consume invalid `spanId` from `Span.INVALID`
+    spanId: '0',
+    traceOptions: types.TraceOptions.UNSAMPLED,
+  });
+
+  /**
+   * Constructs a new Tracer instance.
+   */
+  constructor(config: BasicTracerConfig = {}) {
+    this.sampler = config.sampler || ALWAYS_SAMPLER;
+    this.scopeManager = config.scopeManager;
+    this.defaultAttributes = config.defaultAttributes || {};
+    this.logger = config.logger || new NoopLogger();
+  }
+
+  /**
+   * Starts a new Span or returns the default NoopSpan based on the sampling
+   * decision.
+   */
+  startSpan(name: string, options: types.SpanOptions = {}): types.Span {
+    let parentSpanContext: types.SpanContext | undefined;
+
+    // parent is a SpanContext
+    if (
+      options.parent &&
+      typeof (options.parent as types.SpanContext).traceId
+    ) {
+      parentSpanContext = options.parent as types.SpanContext;
+    }
+    // parent is a Span
+    if (
+      options.parent &&
+      typeof (options.parent as types.Span).context === 'function'
+    ) {
+      parentSpanContext = (options.parent as types.Span).context();
+    }
+
+    // make sampling decision
+    if (!this.sampler.shouldSample(parentSpanContext)) {
+      return BasicTracer.defaultSpan;
```

Comment:
This will not propagate the trace parent and trace state properly.

Reply:
I think even if sampling decision is `false`, we should assign correct `SpanContext` for `DefaultSpan/NoopSpan`? @rochdev this is what you meant, right?

Reply:
I mean that we shouldn't return a single global instance, otherwise it will not be possible to attach the IDs on it for propagation.